src/typechecker.erl

```diff
@@ -3364,51 +3395,75 @@ check_clauses_union(Env, [Ty|Tys], Clauses) ->
-    {VarBinds, constraints:combine(Cs1, Cs2)};
+    {Env1, Cs2} = check_clauses(Env, any, FunResTy, Clauses, bind_vars),
+    {Env1, constraints:combine(Cs1, Cs2)};
 check_clauses_fun(Env, any, Clauses) ->
     check_clauses(Env, any, type(any), Clauses, bind_vars);
 check_clauses_fun(Env, {fun_ty_intersection, Tys, Cs1}, Clauses) ->
-    {VarBinds, Cs2} = check_clauses_intersect(Env, Tys, Clauses),
-    {VarBinds, constraints:combine(Cs1, Cs2)};
+    {Env1, Cs2} = check_clauses_intersect(Env, Tys, Clauses),
+    {Env1, constraints:combine(Cs1, Cs2)};
 check_clauses_fun(Env, {fun_ty_union, Tys, Cs1}, Clauses) ->
-    {VarBinds, Cs2} = check_clauses_union(Env, Tys, Clauses),
-    {VarBinds, constraints:combine(Cs1, Cs2)}.
+    {Env1, Cs2} = check_clauses_union(Env, Tys, Clauses),
+    {Env1, constraints:combine(Cs1, Cs2)}.
 
 %% Checks a list of clauses (if/case/fun/try/catch/receive).
 -spec check_clauses(Env :: #env{}, ArgsTy :: [type()] | any, ResTy :: type(),
                     Clauses :: [gradualizer_type:abstract_clause()],
 		    Caps :: capture_vars | bind_vars) ->
-                        {VarBinds :: map(), constraints:constraints()}.
+                        {VarBinds :: env(), constraints:constraints()}.
 check_clauses(Env, any, ResTy, [{clause, _, Args, _, _} | _] = Clauses, Caps) ->
     %% 'any' is the ... in the type fun((...) -> ResTy)
     ArgsTy = lists:duplicate(length(Args), type(any)),
     check_clauses(Env, ArgsTy, ResTy, Clauses, Caps);
 check_clauses(Env, ArgsTy, ResTy, Clauses, Caps) ->
+    Env1 = push_clauses_controls(Env, #clauses_controls{exhaust = Env#env.exhaust}),
+    %% This is fine, since we match on `any' in the clause above.
+    ArgsTy = ?assert_type(ArgsTy, [type()]),
     %% Clauses for if, case, functions, receive, etc.
-    {VarBindsList, Css, RefinedArgsTy, _VEnvJunk} =
-        lists:foldl(fun (Clause, {VBs, Css, RefinedArgsTy, VEnvIn}) ->
-                            {NewRefinedArgsTy, VB, Cs} =
-                                check_clause(Env#env{venv = VEnvIn},
-                                             RefinedArgsTy, ResTy, Clause, Caps),
-                            VEnvOut =
-                                refine_vars_by_mismatching_clause(Clause, VEnvIn, Env),
-                            {[VB | VBs], [Cs | Css], NewRefinedArgsTy, VEnvOut}
+    {VarBindsList, Css, RefinedArgsTy, Env2} =
+        lists:foldl(fun (Clause, {VBs, Css, RefinedArgsTy, EnvIn}) ->
+                            {NewRefinedArgsTy, Env2, Cs} =
+                                check_clause(EnvIn, RefinedArgsTy, ResTy, Clause, Caps),
+                            VB =
+                                refine_vars_by_mismatching_clause(Clause, Env2#env.venv, Env2),
+                            {[Env2#env{venv = VB} | VBs], [Cs | Css], NewRefinedArgsTy, Env2}
                     end,
-                    {[], [], ArgsTy, Env#env.venv},
+                    {[], [], ArgsTy, Env1},
```

Comment:
I think this is another case where variables bound in one clause survive into the next one, because they're added to Env2 returned by check_clause, which is passed on to the next clause.

Example:

```Erlang
-spec f(integer()) -> integer().
f(X) ->
    case X of
        Y when Y > 42 -> Y;
        Y             -> Y + 1  % a different Y
    end.
```

Solution: Don't add varbinds directly in the returned Env2, alternatively restore venv within Env2 before passing it to the next clause.

Reply:
Hmm, good catch. Thanks!

Reply:
The example above might not show anything in a test case. Maybe this is better, where Y are of different types:

```Erlang
-spec f(foo | integer()) -> integer().
f(X) ->
    case X of
        Y = foo -> length(atom_to_list(Y));
        Y       -> Y + 1  % a different Y
    end.
```

Reply:
Hmm, interesting. I saved this example as a test and was assuming we're done, but...

```erlang
$ cat test/should_pass/variable_binding_leaks.erl
-module(variable_binding_leaks).

-export([case_clauses/1]).

-spec case_clauses(foo | integer()) -> integer().
case_clauses(X) ->
    case X of
        Y = foo -> length(atom_to_list(Y));
        Y       -> Y + 1  % a different Y
    end.
$ cat test/should_pass/type_refinement.erl
-module(type_refinement).

-compile([export_all, nowarn_export_all]).

-spec refine_bound_var_by_guard_bifs(ok | integer() | pid() |
                                     #{x => y} | {x, y}) -> ok.
refine_bound_var_by_guard_bifs(X) ->
    if
        is_map(X)     -> ok;
        is_tuple(X)   -> ok;
        is_integer(X) -> ok;
        is_pid(X)     -> ok;
        true          -> X
    end.
```

The former requires vars to be separate between clauses, the latter depends on refinement by mismatching clause, so the particular var should be updated and propagated 🤷  It needs some more thinking.

Reply:
So currently, we can choose this:

```
$ ./bin/gradualizer test/should_pass/variable_binding_leaks.erl
test/should_pass/variable_binding_leaks.erl: The variable on line 9 at column 9 is expected to have type integer() but it has type foo

    case X of
        Y = foo -> length(atom_to_list(Y));
        Y       -> Y + 1  % a different Y
        ^

$ ./bin/gradualizer test/should_pass/type_refinement.erl
$
```

Or this:

```
$ ./bin/gradualizer test/should_pass/variable_binding_leaks.erl
$ ./bin/gradualizer test/should_pass/type_refinement.erl
test/should_pass/type_refinement.erl: The variable on line 13 at column 26 is expected to have type ok but it has type integer() | ok | #{x => y} | pid() | {x, y}

        is_integer(X) -> ok;
        is_pid(X)     -> ok;
        true          -> X
                         ^

$
```

Reply:
Even more interesting that the current master (3cda2e5) passes both of the above tests 🤔

Reply:
Solved with the today's changes :)